src/ximea_ros_cam.cpp

```diff
@@ -869,9 +767,303 @@ void XimeaROSCam::triggerCb(const std_msgs::Empty::ConstPtr& msg) {
+        }
+    }
+}
+
+void XimeaROSCam::sampleCameraTimestamp(void){
+    uint64_t nanoseconds;
+    ros::Time pre_stamp = ros::Time::now();
+    camera_timestamp_supported_ = get(XI_PRM_TIMESTAMP, nanoseconds) == XI_OK;
+    ros::Time post_stamp = ros::Time::now();
+
+    if(!camera_timestamp_supported_)        
+        return;       
+
+    const uint64_t billion = 1000000000; // c++11 has std::ratio, std::chrono and constexpr for this
+    timestamp_queue_.emplace_back(std::make_pair( pre_stamp, 
+            ros::Duration( nanoseconds/billion, nanoseconds%billion )));
+
+    // ROS_INFO_STREAM("Timestamps:" << // TODO delete or demote to DEGUG
+    //         " System: " << timestamp_queue_.back().first << 
+    //         " Camera: " << timestamp_queue_.back().second );
+
+    // monitor round trip speed
+    ros::Duration round_trip_duration(post_stamp - pre_stamp);
+    if(round_trip_duration > ros::Duration(0.0025)) // arbitary 2.5 ms (I see a 2 ms average with an xiC on my system)
+        ROS_WARN_STREAM_THROTTLE(3, "Camera timestamp query took " << round_trip_duration << " seconds." );    
+}
+
+ros::Time XimeaROSCam::iterpolateTimestamp(const XI_IMG& frame){
+    if(!camera_timestamp_supported_)  
+        return ros::Time::now();
+
+    sampleCameraTimestamp();
+
+    std::deque< std::pair<ros::Time,ros::Duration> >& queue = timestamp_queue_; // alias 
+    if(queue.size() < 2)
+    {  
+        ROS_ERROR("Not enough timestamp samples to interpolate.");
+        return ros::Time::now();
+    }
+
+    ros::Duration camera_stamp(frame.tsSec, frame.tsUSec * 1000);
+    // TODO adjust camera stamp to be the start of exposure (model dependant)
+    // consider using a precalculated bool to determin if exposure time should
+    // be deducted  
+    
+    bool clock_wrap = queue.front().second > queue.back().second; 
+    if(clock_wrap)
+    {
+        ROS_WARN("Clock wrap detected. Timestamp handling experimetal!");
+        // xiQ, xiD 
+        // 40 bit microsecond number - (overlaps after 305 hours)
+        // TimeStamp is recorded at the start of Data Readout 
+
+        // xiC, xiB, xiX, xiT 
+        // 64 bit 4-nanoseconds number (overlaps after 2339 years) (sic)
+        // TimeStamp is recorded at the start of Exposure.
+        // It is assumed that these cameras use the full tsSec value 
+
+        // untested experimental handling!
+        ros::Duration wrap_period;
+        ros::Duration rough_wrap_period = 
+                (queue.front().second - queue.back().second) + 
+                (queue.back().first - queue.front().first);
+
+        const uint64_t million = 1000000;
+        const uint64_t forty_bits = uint64_t(1) << 40;
+        const ros::Duration forty_bit_period(forty_bits / million, forty_bits % million);
+        const ros::Duration ros_duration_period(1<<31,0); // ros duration overflow
+
+        if((rough_wrap_period - forty_bit_period).sec == 0) 
```

Comment:
No (I hope) because we account for the system time difference so I believe the error will be the drift between the camera and system clocks together with the round trip error which I imagine to be in the order of milliseconds. It really needs testing in some way though. I'm thinking of artificially simulating it rather than waiting for 12 days.